## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
df.describe

<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.max()


first                              Zachary
last                                Zavala
job                           Youth worker
phone                    999.801.2159x2255
email          zacharyjohnson@jacobson.net
client_id                            99984
order_id                           9998480
order_date                      2023-05-30
order_week                              52
order_year                            2023
item_id                     ZZU43604-44-9M
category                          software
subcategory                       wall art
unit_price                         1396.23
unit_cost                           846.27
unit_weight                          46.43
qty                                3958244
line_number                              9
dtype: object

In [5]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print("Top 3 categories:/n",top_categories)

Top 3 categories:/n category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [6]:
# For the category with the most entries, which subcategory had the most entries?
top_category = top_categories.index[0]
top_subcategory = df[df['category']==top_category]['subcategory'].value_counts().idxmax()

In [7]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print("top 5 clients:\n", top_clients)

top 5 clients:
 client_id
33615    220
66037    211
46820    209
24741    207
38378    207
Name: count, dtype: int64


In [8]:
# Store the client ids of those top 5 clients in a list.
top_clients_id = top_clients.index.tolist()
print("top 5 clients ids:\n",top_clients_id) 

top 5 clients ids:
 [33615, 66037, 46820, 24741, 38378]


In [9]:
# How many total units (the qty column) did the client with the most entries order order?
top_clients_id = top_clients_id[0]
total_units = df[df['client_id']==top_clients_id]['qty'].sum()
print("total units ordered by client with most entries:\n",total_units)

total units ordered by client with most entries:
 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']


In [11]:
df['total_weight'] = df['unit_weight'] * df['qty']

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10 *x)
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'line_subtotal', 'total_weight', 'shipping_price'],
      dtype='object')


In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df['line_subtotal'] + df['shipping_price']
df['total_price'] *= 1.09249997

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df['profit'] = df['total_price'] - df['line_cost']

In [15]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
orders = [2742071,  2173913, 6128929]

for order_id in orders: 
    order_df= df[df['order_id']==order_id]
    total_price = order_df['total_price'].sum()
    print(f"Order Id {order_id} Total Price: ${total_price:.2f}")

Order Id 2742071 Total Price: $152811.89
Order Id 2173913 Total Price: $162388.71
Order Id 6128929 Total Price: $923441.22


In [17]:
top_clients = df['client_id'].value_counts().head(5)
top_clients_id = top_clients.index.tolist()
top_clients_id

[33615, 66037, 46820, 24741, 38378]

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [18]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
spending_summary = df[df['client_id'].isin(top_clients_id)].groupby('client_id').agg({
    'total_price': 'sum'
}).reset_index()
print("Total Spending by Top 5 Clients:\n", spending_summary)

Total Spending by Top 5 Clients:
    client_id   total_price
0      24741  8.226889e+07
1      33615  8.377308e+06
2      38378  1.290655e+07
3      46820  9.743794e+06
4      66037  1.025951e+07


In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
spending_summary = df[df['client_id'].isin(top_clients_id)].groupby('client_id').agg({
    'total_price':'sum'
}).reset_index()
print("total spending by top 5 clients:\n",spending_summary)


total spending by top 5 clients:
    client_id   total_price
0      24741  8.226889e+07
1      33615  8.377308e+06
2      38378  1.290655e+07
3      46820  9.743794e+06
4      66037  1.025951e+07


In [20]:
#f = of top 5 clients
#cols = ['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']
#for i in cols:
   # df[i] = [totals_foryourclentsdf(client_id, col) for client_id in top_5_clientsdf]
#summary_df = pd.DataFrame(df)
top_df = df[df['client_id'].isin(top_clients_id)]
#top_df

In [21]:
print(top_df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'line_subtotal', 'total_weight', 'shipping_price',
       'total_price', 'line_cost', 'profit', 'line_profit'],
      dtype='object')


In [22]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_df_organized = top_df[['client_id', 
                                         'qty', 
                                         'shipping_price',
                                         'total_price',
                                         'line_cost',
                                         'line_profit'
                                         ]]
top_df_organized

client_id  qty  shipping_price    total_price  line_cost   line_profit
5          33615   78         1523.34   16392.809100   13569.66   2823.149100
12         33615  568       109379.76  262060.240204  222059.60  40000.640204
137        46820  118         4666.90   22647.786578   17820.36   4827.426578
214        46820  133         6563.55    8330.192096    7396.13    934.062096
216        66037  157          408.20   52223.454141   26484.33  25739.124141
...          ...  ...             ...            ...        ...           ...
54535      66037  108         5685.12    8188.505775    6836.40   1352.105775
54536      24741  721        23720.90   76185.616508   52596.95  23588.666508
54579      38378  118        10795.82   15823.026666   13190.04   2632.986666
54589      38378  685         3356.50   42761.432076   29564.60  13196.832076
54613      24741   91         8363.81   18648.734138   13761.93   4886.804138

[1054 rows x 6 columns]

In [23]:
top_df_compresed = top_df_organized.groupby(['client_id'], as_index=False)[['qty','shipping_price','total_price','line_cost','line_profit']].sum()

In [24]:
top_df_compresed

client_id     qty  shipping_price   total_price    line_cost   line_profit
0      24741  239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07
1      33615   64313      1828984.89  8.377308e+06   6175313.91  2.201994e+06
2      38378   73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
3      46820   75768      1601448.84  9.743794e+06   7007191.64  2.736602e+06
4      66037   43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06

In [25]:
# Format the data and rename the columns to names suitable for presentation.
top_df_renamed = top_df_compresed.rename(columns ={'client_id' : 'Client ID',
                                                    'qty' : 'Units',
                                                    'shipping_price' : 'Shipping',
                                                    'total_price' : 'Total Revenue',
                                                    'line_cost' : 'Total Cost',
                                                    'line_profit' : 'Total Profit'
                                                   })


# Define the money columns. 
money_columns = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
    return amount / 1_000_000 

for column in money_columns:
    top_df_renamed[column]= top_df_renamed[column].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top_df_new = top_df_renamed.rename(columns ={'Shipping' : 'Shipping (millions)',
                                            'Total Revenue' : 'Total Revenue (millions)',
                                            'Total Cost' : 'Total Cost (millions)',
                                            'Total Profit' : 'Total Profit (millions)'
                                            })



In [28]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
top_df_new = top_df_new.sort_values('Total Profit (millions)',ascending=False)
top_df_sort = top_df_new.reset_index(drop=True) 

top_df_sort

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268890   
1      38378   73667             3.429455                 12.906551   
2      66037   43018             1.395152                 10.259515   
3      46820   75768             1.601449                  9.743794   
4      33615   64313             1.828985                  8.377308   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579990  
1               9.634721                 3.271830  
2               7.004483                 3.255032  
3               7.007192                 2.736602  
4               6.175314                 2.201994